src/Dialect/ONNX/Rewrite.cpp

```diff
@@ -312,30 +311,87 @@ class LoopOpRewriteMaxTripCountPattern : public OpRewritePattern<ONNXLoopOp> {
-      stepValue = getFedValue(stepValue, loopOp);
-    if (!isDefinedByIntegerConstantOp(stepValue))
-      return std::make_pair(false, -1);
-    // 2. Counter is an block argument and updated at each iteration.
+    // Counter is an block argument and updated at each iteration.
     if (!isUpdatedArgByValue(counterValue, newCounterValue, returnOp))
-      return std::make_pair(false, -1);
-    // 3. Counter is initially fed by a constant.
-    Value startValue = getFedValue(counterValue, loopOp);
-    if (!isDefinedByIntegerConstantOp(startValue))
-      return std::make_pair(false, -1);
-
-    // Check that the new trip count is smaller than the original trip count.
-    int64_t lowerBound = getOneIntergerConstant(startValue);
-    int64_t upperBound = getOneIntergerConstant(ubValue);
-    int64_t step = getOneIntergerConstant(stepValue);
-    if ((step <= 0) || (upperBound - lowerBound) <= 0)
-      return std::make_pair(false, -1);
-    int64_t derivedTripCount =
-        ceil((1.0 * (upperBound - lowerBound)) / (1.0 * step));
-    int64_t maxTripCount = getOneIntergerConstant(maxTripCountValue);
-
-    return std::make_pair(maxTripCount > derivedTripCount, derivedTripCount);
+      return std::make_pair(false, maxTripCountValue);
+    // Step must be a constant inside the loop or an invariant argument.
+    if (!isIntConstantOrInvariantBlockArg(stepValue, returnOp))
+      return std::make_pair(false, maxTripCountValue);
+
+    // Check the lower bound of the break condition.
+    // LowerBound is the initial value of the counter.
+    Value lbValue = getFedValue(counterValue, loopOp);
+
+    // Check the upper bound of the break condition.
+    // UpperBound must be a constant inside the loop or an invariant argument.
+    if (!isIntConstantOrInvariantBlockArg(ubValue, returnOp))
+      return std::make_pair(false, maxTripCountValue);
+
+    // Get values for upper bound and step if they are invariant arguments.
+    // Otherwise, clone them to location outside the loop.
+    if (isInvariantBlockArg(ubValue, returnOp))
+      ubValue = getFedValue(ubValue, loopOp);
+    else
+      ubValue = cast<ONNXConstantOp>(rewriter.clone(*ubValue.getDefiningOp()))
+                    .getResult();
+    if (isInvariantBlockArg(stepValue, returnOp))
+      stepValue = getFedValue(stepValue, loopOp);
+    else
+      stepValue =
+          cast<ONNXConstantOp>(rewriter.clone(*stepValue.getDefiningOp()))
+              .getResult();
+
+    // Case 1: the upper bound, lower bound and step are constants.
+    // - Compute the new max trip count at the compile time.
+    if (isDefinedByIntegerConstantOp(lbValue) &&
+        isDefinedByIntegerConstantOp(ubValue) &&
+        isDefinedByIntegerConstantOp(stepValue)) {
+      int64_t lowerBound = getOneIntergerConstant(lbValue);
+      int64_t upperBound = getOneIntergerConstant(ubValue);
+      int64_t step = getOneIntergerConstant(stepValue);
+      if ((step <= 0) || (upperBound - lowerBound) <= 0)
```

Comment:
nit `upperBound <= lowerBound` is easier to read. And if the trip count is zero? we could just get rid of the loop :-)